src/driver/hermes_cdev.c

```diff
@@ -213,13 +213,66 @@ static long hermes_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 	}
 }
 
+static ssize_t hermes_read_iter(struct kiocb *iocb, struct iov_iter *to)
+{
+	int i;
+	ssize_t res = 0, bytes;
+	loff_t offset = iocb->ki_pos;
+
+	if (iocb->ki_flags & (IOCB_HIPRI | IOCB_NOWAIT))
+		return -EOPNOTSUPP;
+
+	if (offset == -1)
+		return -EOPNOTSUPP;
+
+	for (i = 0; i < to->nr_segs; i++) {
+		bytes = hermes_read(iocb->ki_filp, to->iov[i].iov_base,
+				to->iov[i].iov_len, &offset);
```

Comment:
No, read_iter() should be considered as a single read, not multiple reads.

hermes_read_iter() should replace hermes_read()... you don't need both and it should do exactly what hermes_read() does, just with different input arguments.

Reply:
Hmm, yes, that makes sense. I'll see how to put everything into a single xdma request.

Reply:
It doesn't necessarily have to be a single XDMA request....

But all the setup and job execution stuff has to be done only once.

Reply:
What do you mean with the job execution?

`hermes_read()` is already pretty concise. Certainly the `if (env->data_slot < 0)` can be pulled out and done only once (and maybe the `xdma_get_c2h()` call too), but the rest needs to be done for each iovec.

Reply:
Hmm, yes well I mean this to apply to both read_iter and write_iter..... Part of it is I didn't bother to look at those functions and only saw how the new function looked.

But I'd expect read_iter() to simply be what read() is now with a loop around the xdma_channel_read_write().... Same for write_iter(). I'd structure the code and resulting patchset with that in mind. So instead of creating a new read_iter() function that uses the old read() function I'd modify the old function to be read_iter().

Also an unrelated nit for future cleanup: I'd make a couple of static inline wrappers for xdma_channel_read_write() one for xdma_channel_read() and one for xdma_channel_write()... makes the code a little more readable and less error prone to not have to know what the boolean polarity is.

Reply:
Gotcha, thanks. I'll fix those and resubmit